File: BMWCenter/Core/Util/Formatters.swift
import Foundation

enum Formatters {
    private static let durationComponents: DateComponentsFormatter = {
        let f = DateComponentsFormatter()
        f.allowedUnits = [.hour, .minute]
        f.unitsStyle = .abbreviated
        f.zeroFormattingBehavior = .dropLeading
        return f
    }()

    private static let liveDuration: DateComponentsFormatter = {
        let f = DateComponentsFormatter()
        f.allowedUnits = [.hour, .minute, .second]
        f.unitsStyle = .positional
        f.zeroFormattingBehavior = [.pad]
        return f
    }()

    private static let currency: NumberFormatter = {
        let f = NumberFormatter()
        f.numberStyle = .currency
        return f
    }()

    static let unavailable = String(localized: "common.value.unavailable", table: "Localizable")

    static func number(_ value: Double?, digits: Int = 1) -> String {
        guard let value else { return unavailable }
        return MetricFormatter.number(value, fractionLength: digits)
    }

    static func speed(_ kmh: Double?, settings: AppSettings) -> String {
        guard let kmh else { return unavailable }
        let value: Double
        let unit: String
        switch settings.unitSystem {
        case .metric:
            value = kmh
            unit = String(localized: "unit.kmh", table: "Localizable")
        case .imperial:
            value = kmh * 0.621371
            unit = String(localized: "unit.mph", table: "Localizable")
        }
        return "\(number(value, digits: 0)) \(unit)"
    }

    static func distance(_ km: Double?, settings: AppSettings) -> String {
        guard let km else { return unavailable }
        switch settings.unitSystem {
        case .metric:
            return "\(number(km, digits: 1)) km"
        case .imperial:
            return "\(number(km * 0.621371, digits: 1)) mi"
        }
    }

    static func temperature(_ celsius: Double?, settings: AppSettings) -> String {
        guard let celsius else { return unavailable }
        switch settings.temperatureUnit {
        case .celsius:
            return "\(number(celsius, digits: 0)) \(String(localized: "unit.celsius", table: "Localizable"))"
        case .fahrenheit:
            let f = celsius * 9 / 5 + 32
            return "\(number(f, digits: 0)) \(String(localized: "unit.fahrenheit", table: "Localizable"))"
        }
    }

    static func rpm(_ value: Double?) -> String {
        guard let value else { return unavailable }
        return "\(number(value, digits: 0)) \(String(localized: "unit.rpm", table: "Localizable"))"
    }

    static func percent(_ value: Double?) -> String {
        guard let value else { return unavailable }
        return "\(number(value, digits: 0))\(String(localized: "unit.percent", table: "Localizable"))"
    }

    static func voltage(_ value: Double?) -> String {
        guard let value else { return unavailable }
        return "\(number(value, digits: 1)) \(String(localized: "unit.volt", table: "Localizable"))"
    }

    static func boost(_ bar: Double?, settings: AppSettings) -> String {
        guard let bar else { return unavailable }
        switch settings.pressureUnit {
        case .bar:
            let sign = bar >= 0 ? "+" : ""
            return "\(sign)\(number(bar, digits: 2)) bar"
        case .kpa:
            return "\(number(bar * 100, digits: 0)) kPa"
        case .psi:
            return "\(number(bar * 100 * 0.145038, digits: 1)) psi"
        }
    }

    static func liters(_ value: Double?) -> String {
        guard let value else { return unavailable }
        return "\(number(value, digits: 2)) \(String(localized: "unit.liter", table: "Localizable"))"
    }

    /// Convert L/100km to display string per settings.
    static func consumption(l100km: Double?, idleLh: Double? = nil, speedKmh: Double? = nil, settings: AppSettings) -> String {
        if let speed = speedKmh, speed <= 3, let idleLh {
            return "\(number(idleLh, digits: 1)) \(String(localized: "unit.literPerHour", table: "Localizable"))"
        }
        guard let l100km, l100km >= 0.5, l100km <= 60 else { return unavailable }
        let value: Double
        let unitKey: String
        switch settings.consumptionUnit {
        case .l100km:
            value = l100km
            unitKey = "unit.l100km"
        case .kmPerL:
            value = 100 / l100km
            unitKey = "unit.kmPerL"
        case .mpgUS:
            value = 235.215 / l100km
            unitKey = "unit.mpgUS"
        case .mpgUK:
            value = 282.481 / l100km
            unitKey = "unit.mpgUK"
        }
        return "\(number(value, digits: 1)) \(String(localized: String.LocalizationValue(unitKey), table: "Localizable"))"
    }

    static func duration(_ seconds: Double) -> String {
        if seconds < 3600 {
            let mins = Int((seconds / 60).rounded())
            return durationComponents.string(from: DateComponents(minute: max(mins, 0))) ?? unavailable
        }
        return durationComponents.string(from: seconds) ?? unavailable
    }

    static func liveDuration(_ seconds: Double) -> String {
        liveDuration.string(from: max(0, seconds)) ?? "00:00:00"
    }

    static func currency(_ amount: Double, code: String) -> String {
        currency.currencyCode = code
        return currency.string(from: NSNumber(value: amount)) ?? number(amount, digits: 2)
    }

    static func truncate(_ text: String, max: Int) -> String {
        guard text.count > max else { return text }
        return String(text.prefix(max - 1)) + "…"
    }
}

File: mobile/src/design/formatters.ts
import i18n from "@/i18n";
import { MetricFormatter, formatNumber } from "./metricFormatter";
import type { AppSettingsState } from "@/core/settings/appSettings";

type Settings = Pick<
  AppSettingsState,
  "unitSystem" | "temperatureUnit" | "consumptionUnit" | "pressureUnit"
>;

const t = (key: string) => i18n.t(key);

export const unavailable = () => t("common.value.unavailable");

export const Formatters = {
  number(value: number | undefined, digits = 1): string {
    if (value == null) return unavailable();
    return formatNumber(value, digits);
  },

  speed(kmh: number | undefined, settings: Settings): string {
    if (kmh == null) return unavailable();
    const value = settings.unitSystem === "metric" ? kmh : kmh * 0.621371;
    const unit = settings.unitSystem === "metric" ? t("unit.kmh") : t("unit.mph");
    return `${this.number(value, 0)} ${unit}`;
  },

  distance(km: number | undefined, settings: Settings): string {
    if (km == null) return unavailable();
    if (settings.unitSystem === "metric") return `${this.number(km, 1)} km`;
    return `${this.number(km * 0.621371, 1)} mi`;
  },

  /** Odometer and service intervals are whole numbers — a tenth of a kilometre
   * is noise next to a six-figure reading. */
  odometer(km: number | undefined, settings: Settings): string {
    if (km == null) return unavailable();
    if (settings.unitSystem === "metric") return `${this.number(km, 0)} km`;
    return `${this.number(km * 0.621371, 0)} mi`;
  },

  temperature(celsius: number | undefined, settings: Settings): string {
    if (celsius == null) return unavailable();
    if (settings.temperatureUnit === "celsius") return `${this.number(celsius, 0)} ${t("unit.celsius")}`;
    const f = (celsius * 9) / 5 + 32;
    return `${this.number(f, 0)} ${t("unit.fahrenheit")}`;
  },

  rpm(value: number | undefined): string {
    if (value == null) return unavailable();
    return `${this.number(value, 0)} ${t("unit.rpm")}`;
  },

  percent(value: number | undefined): string {
    if (value == null) return unavailable();
    return `${this.number(value, 0)}${t("unit.percent")}`;
  },

  voltage(value: number | undefined): string {
    if (value == null) return unavailable();
    return `${this.number(value, 1)} ${t("unit.volt")}`;
  },

  boost(bar: number | undefined, settings: Settings): string {
    if (bar == null) return unavailable();
    switch (settings.pressureUnit) {
      case "bar": {
        const sign = bar >= 0 ? "+" : "";
        return `${sign}${this.number(bar, 2)} bar`;
      }
      case "kpa":
        return `${this.number(bar * 100, 0)} kPa`;
      case "psi":
        return `${this.number(bar * 100 * 0.145038, 1)} psi`;
    }
  },

  liters(value: number | undefined): string {
    if (value == null) return unavailable();
    return `${this.number(value, 2)} ${t("unit.liter")}`;
  },

  /** Convert L/100km to display string per settings. */
  consumption(l100km: number | undefined, settings: Settings, idleLh?: number, speedKmh?: number): string {
    if (speedKmh != null && speedKmh <= 3 && idleLh != null) {
      return `${this.number(idleLh, 1)} ${t("unit.literPerHour")}`;
    }
    if (l100km == null || l100km < 0.5 || l100km > 60) return unavailable();
    let value: number;
    let unitKey: string;
    switch (settings.consumptionUnit) {
      case "l100km":
        value = l100km;
        unitKey = "unit.l100km";
        break;
      case "kmPerL":
        value = 100 / l100km;
        unitKey = "unit.kmPerL";
        break;
      case "mpgUS":
        value = 235.215 / l100km;
        unitKey = "unit.mpgUS";
        break;
      case "mpgUK":
        value = 282.481 / l100km;
        unitKey = "unit.mpgUK";
        break;
    }
    return `${this.number(value, 1)} ${t(unitKey)}`;
  },

  duration(seconds: number): string {
    if (seconds < 3600) {
      const mins = Math.max(0, Math.round(seconds / 60));
      return `${mins}m`;
    }
    const hours = Math.floor(seconds / 3600);
    const mins = Math.round((seconds % 3600) / 60);
    return `${hours}h ${mins}m`;
  },

  liveDuration: MetricFormatter.liveDuration,

  currency(amount: number, code: string): string {
    try {
      return new Intl.NumberFormat(i18n.language, { style: "currency", currency: code }).format(amount);
    } catch {
      return `${formatNumber(amount, 2)} ${code}`;
    }
  },

  truncate(text: string, max: number): string {
    if (text.length <= max) return text;
    return `${text.slice(0, max - 1)}…`;
  },
};
